src/tir/ir/stmt.cc

```diff
@@ -274,9 +274,10 @@ TVM_STATIC_IR_FUNCTOR(ReprPrinter, vtable)
 // Allocate
 Allocate::Allocate(Var buffer_var, DataType dtype, Array<PrimExpr> extents, PrimExpr condition,
                    Stmt body, Span span) {
-  // TODO(tvm-team): Add invariant check to make sure
-  // IsPointerPType(buffer_var->type_annotation, dtype)
-  // once we fix the allocate tvm script printing.
+  ICHECK(IsPointerType(buffer_var->type_annotation, dtype))
+      << "Allocate: buffer_var expect to have the right pointer type annotation"
+      << " annotation=" << buffer_var->type_annotation << ", dtype=" << dtype;
```

Comment:
This might be a better error message:
```suggestion
      << "The allocated data type (" << dtype << ") does not match the type annotation of the buffer " << buffer_var << " (" << buffer_var->type_annotation << "). The data type should be an element of the pointer type."
```

It does seem like we could infer the type from the buffer_var though...

Reply:
This should probably be a `CHECK` too if people are writing tvmscript.